specifyweb/frontend/js_src/lib/components/InitialContext/treeRanks.ts

```diff
@@ -116,22 +104,65 @@ function getTreeScope(
   );
 }
 
+export function getTreeDefinitions<TREE_NAME extends AnyTree['tableName']>(
+  tableName: TREE_NAME,
+  definitionFilter?: Partial<
+    SerializedResource<FilterTablesByEndsWith<'TreeDef'>>
+  >
+): typeof treeDefinitions[TREE_NAME] {
+  const specificTreeDefinitions = caseInsensitiveHash(
+    defined(treeDefinitions),
+    tableName
+  );
+
+  return definitionFilter === undefined
+    ? specificTreeDefinitions
+    : specificTreeDefinitions.filter(({ definition }) =>
+        Object.entries(definitionFilter).every(
+          ([key, value]) => caseInsensitiveHash(definition, key) === value
+        )
+      );
+}
+
 export function getTreeDefinitionItems<TREE_NAME extends AnyTree['tableName']>(
   tableName: TREE_NAME,
-  includeRoot: boolean
-): typeof treeDefinitions[TREE_NAME]['ranks'] | undefined {
-  const definition = caseInsensitiveHash(treeDefinitions, tableName);
-  return definition?.ranks.slice(includeRoot ? 0 : 1);
+  includeRoot: boolean,
+  definitionFilter?: Partial<
+    SerializedResource<FilterTablesByEndsWith<'TreeDef'>>
+  >
```

Comment:
idea: instead of this filter could we simply have `treeDefinitionId: number | undefined`?

pros:
- it's requires so this would avoid bugs when the caller of this function forgets to provide it
- simplifies the code - we can add back complexity when if it would be needed

same goes for getTreeDefinitions

Reply:
even better, you can have "getCache('tree', `definition${tableName}`)," be the default value for treeDefinitionId, thus making the default "just work"

Reply:
> idea: instead of this filter could we simply have treeDefinitionId: number | undefined?

yes, I agree in nearly cases just `treeDefinitionId: number | undefined` would be better. And in all uses, we do just use the id. 

I had assumed that some point in the future we may need to return a list of certain TreeDefItems or TreeDefs fitting some criteria, but that is probably not that likely (and the caller can do so anyways if needed)

Reply:
> even better, you can have "getCache('tree', definition${tableName})," be the default value for treeDefinitionId, thus making the default "just work"

Did you have something like this in mind? 
https://github.com/specify/specify7/pull/5036/commits/82fb29d8477e64e2b9c659a86c3258ff22549ceb

Reply:
looks great! that reduced a bit of complexity